### PR TITLE
Added serialization phase

### LINKED ISSUE #603: New Plugin Phase: Serialization
## Detailed Description

Introduce a new plugin phase *serialization* (or *request-serialization*?) to allow plugins to post-process serialized request bodies.

## Context

Enables the following use cases:
- #608 `RequestCompressionPlugin` (rather than an interceptor that is specific to Apache HTTP Client)
- `LogbookPlugin` (rather than an interceptor)
- Allows us to properly implement the plugin interceptor
  - Pass serialized body to the right place

https://github.com/zalando/riptide/blob/ad63d736a1daff1690eed1e496b8c39c43a887fc/riptide-core/src/main/java/org/zalando/riptide/PluginInterceptor.java#L118-L130

## Possible Implementation

- [x] Phase *serialization*, `Plugin.aroundSerialization`
- [x] After *dispatch*, but before *network*
- [x] Implementation: Delegate to `MessageWriter`
- [x] Keep original and serialized body in `RequestArguments`
  - [x] `body`: Java bean?
  - [x] `entity`: byte array?

### PR DESCRIPTION
**Pull request by whiskeysierra**
## Description

Requires #607 

## Motivation and Context

Fixes #603 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: riptide-core/src/main/java/org/zalando/riptide/SerializationPlugin.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            final HttpHeaders headers = new HttpHeaders();
+            headers.addAll(toMultiValueMap(arguments.getHeaders()));
+            final ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            final HttpOutputMessage output = new CapturingHttpOutputMessage(headers, stream);
+
+            writer.write(output, arguments);
+
+            return execution.execute(arguments
+                    .replaceHeaders(headers)
+                    .withEntity(stream.toByteArray()));
```
Comment (lukasniemeier-zalando) [inline riptide-core/src/main/java/org/zalando/riptide/SerializationPlugin.java]: I wonder how often we now copy the request body. Two times or three times?

**Comment by lukasniemeier-zalando** [comment]
:+1:

File: riptide-core/src/main/java/org/zalando/riptide/SerializationPlugin.java (hunk 1 above)
Comment (whiskeysierra) [reply riptide-core/src/main/java/org/zalando/riptide/SerializationPlugin.java]: This should be the first time within Riptide now. I'm thinking about using an `OutputStream` instead but it's certainly harder to use then.

Comment (whiskeysierra) [reply riptide-core/src/main/java/org/zalando/riptide/SerializationPlugin.java]: Logbook would do another copy, unless we implement it natively within Riptide #604 and re-use the copy of the entity.

File: riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/GzipHttpRequestInterceptor.java [hunk 2]
```diff
@@ -34,7 +35,7 @@ public void process(final HttpRequest original, final HttpContext context) {
 
     private boolean hasNonEmptyBody(final HttpEntityEnclosingRequest request) {
         return Optional.ofNullable(request.getEntity())
```
Comment (lukasniemeier-zalando) [inline riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/GzipHttpRequestInterceptor.java]: just `request.getEntity() != null`?

**Comment by lukasniemeier-zalando** [comment]
:+1:

File: riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/GzipHttpRequestInterceptor.java (hunk 2 above)
Comment (whiskeysierra) [reply riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/GzipHttpRequestInterceptor.java]: :man_facepalming: yes...

**Pull request merged**